lib/lambda_ethereum_consensus/store/state_store.ex

```diff
@@ -6,61 +6,80 @@ defmodule LambdaEthereumConsensus.Store.StateStore do
   alias LambdaEthereumConsensus.Store.Utils
   alias Types.BeaconState
 
-  @state_prefix "beacon_state"
-  @stateslot_prefix @state_prefix <> "slot"
+  @state_prefix_by_block "beacon_state_by_block"
+  @state_prefix_by_state "beacon_state_by_state"
+  @stateslot_prefix_by_block @state_prefix_by_block <> "slot"
+  @stateslot_prefix_by_state @state_prefix_by_state <> "slot"
 
   @spec store_state(BeaconState.t()) :: :ok
   def store_state(%BeaconState{} = state) do
     # NOTE: due to how SSZ-hashing works, hash(block) == hash(header)
     block_root = Ssz.hash_tree_root!(state.latest_block_header)
+    state_root = Ssz.hash_tree_root!(state)
     {:ok, encoded_state} = Ssz.to_ssz(state)
 
-    key = state_key(block_root)
-    Db.put(key, encoded_state)
+    key_block = state_key(block_root, :block_root)
+    key_state = state_key(state_root, :state_root)
```

Comment:
States are really heavy, so ideally we'd have them on one place only. I auggest just leaving the old block_hash -> state mapping (that way we don't impact block processing), and just add a state_hash -> block_hash mapping. Then when fetching the state by state hash we can just do state_hash -> block_hash -> state